packages/jupyter-ai/jupyter_ai/tools/models.py

```diff
@@ -85,35 +86,36 @@ def set_name_description(self):
 
         return self
 
+    @property
+    def is_async(self):
+        """
+        Returns whether this tool's `callable` is an async function.
+        """
+        return inspect.iscoroutinefunction(self.callable)
+
     def __eq__(self, other):
         if not isinstance(other, Tool):
             return False
-        return self.name == other.name
+        return (
+            self.name == other.name
+            and self.description == other.description
+            and self.read == other.read
+            and self.write == other.write
+            and self.delete == other.delete
+            and self.execute == other.execute
+        )
 
     def __hash__(self):
-        return hash(self.name)
-
-
-class ToolSet(set):
-    """A collection of tools that enforces unique tool names.
-
-    Ensures that tools with duplicate names cannot be added to the same collection.
-
-    Example:
-        >>> toolset = ToolSet()
-        >>> tool1 = Tool(callable=lambda: None, name="test")
-        >>> tool2 = Tool(callable=lambda: None, name="test")  # Same name
-        >>> toolset.add(tool1)
-        >>> toolset.add(tool2)  # Raises ValueError
-        Traceback (most recent call last):
-        ...
-        ValueError: Tool with name 'test' already exists in the set
-    """
-
-    def add(self, item):
-        if item in self:
-            raise ValueError(f"Tool with name '{item.name}' already exists in the set")
-        super().add(item)
+        return hash(
+            (
+                self.name,
+                self.description,
+                self.read,
+                self.write,
+                self.delete,
+                self.execute,
+            )
+        )
```

Comment:
This will cause a collision for tools that share the same name, description, and capabilities even if they call different functions. It may be better to avoid collisions entirely by hashing the memory address of the callable function:

```suggestion
        return hash(id(self.callable))
```

This ensures:

1. Tools that reference the same function produce the same hash. This could allow us to detect duplicate tools and ignore them if necessary.
2. Tools that have different callables will always be distinguishable, even if every other attribute is the same. Could be helpful in dev/testing scenarios.